pkg/storegateway/indexheader/encoding/reader.go

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package encoding
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+)
+
+type FileReader struct {
+	file   *os.File
+	buf    *bufio.Reader
+	base   int
+	length int
+	pos    int
+}
+
+// NewFileReader creates a new FileReader for the segment of file beginning at base
+// with length total length using the supplied buffered reader.
+func NewFileReader(file *os.File, base, length int, buf *bufio.Reader) (*FileReader, error) {
+	f := &FileReader{
+		file:   file,
+		buf:    buf,
+		base:   base,
+		length: length,
+	}
+
+	err := f.Reset()
+	if err != nil {
+		return nil, err
+	}
+
+	return f, nil
+}
+
+// Reset moves the cursor position to the beginning of the file segment.
+func (f *FileReader) Reset() error {
+	return f.ResetAt(0)
+}
+
+// ResetAt moves the cursor position to the given absolute offset in the file segment.
+// Attempting to ResetAt to the end of the file is valid. Attempting to ResetAt _beyond_
+// the end of the file will return an error.
+func (f *FileReader) ResetAt(off int) error {
+	if off > f.length {
+		return ErrInvalidSize
+	}
+
+	pos, err := f.file.Seek(int64(f.base+off), io.SeekStart)
+	if err != nil {
+		return err
+	}
+
+	f.buf.Reset(f.file)
+	f.pos = int(pos) - f.base
```

Comment:
Looks a bit convoluted. Isn't `f.pos = off`?

Reply:
Good point, seems like it. I'll let @charleskorn take a look here since he added the `f.pos` functionality.

Reply:
Yep, you're right, it should be equivalent to `off`. Resolved in a76db59.